### ISSUE TITLE
RFC: graphene invocation

### ISSUE THREAD

**Issue by woju** [oscarlab/graphene#1053]
# Introduction

I'd like to propose a unified way of running files in graphene. This will certainly be a very visible, very incompatible change, so I'd like to write some explanation and seek review, so we can get this right. Also, I'm not that knowledgeable about the architecture, so I apologize in advance if i had some misconceptions.

### Requirements
- command line invocations (hard)
- invocation from shebang's manifest (nice to have)
- easy way of running under GDB (nice to have, developers love it, so probably a must)

### Status quo ante
Today it is mainly done with `Runtime/pal_loader` script. Script 

### Related work
- #895 
- that CMake/meson/whatever effort to rewrite buildsystem, which I can't find the issue number for
- #994 

# the proposal (DRAFT)

After proper installation (`apt-get install` or `make DESTDIR=/ install`), `/usr/bin/graphene` and `/usr/bin/graphene-sgx` are binaries or symlinks. No `Linux` or `FreeBSD`, since user knows what the `uname` is or should be. In both cases the interface is the same:

1. `graphene[-sgx] [--gdb] {--manifest|-f} MANIFEST`, and manifest contains path to executable and argv
2. `graphene[-sgx] [--gdb] {--manifest|-f} MANIFEST [--] EXECUTABLE [ARGS ...]`, and the manifest contains neither executable nor argv
3. `graphene[-sgx] [--gdb] [--] EXECUTABLE [ARGS]` for the cases the manifest is not needed

This makes it possible to use it from manifest's shebang like awk: `#!/usr/bin/graphene -f`. If the manifest is invoked directly (`./manifest`), it is not possible to automatically run GDB, but as always, it is possible to use `graphene --gdb -f ./manifest`.

It is an error to specify arguments twice, both in command line and in manifest, even if they match.

Environment variables are not accepted.

# potential problems

### shebangs under FreeBSD
*BSD people don't like putting alien binaries in `/usr/bin`, and would likely prefer to have those binaries installed in `/usr/local/bin`. This will break shebangs as proposed. The Python way of writing `#!/usr/bin/env python3` (`#!/usr/bin/env graphene -f`) would not work, since `graphene -f` would be a single argument to `env` (not two arguments: `graphene` and `-f`). One solution would be to follow the Perl way of installing a symlink `/usr/bin/graphene` → `/usr/local/bin/graphene`.

# to consider
Maybe `--gdb` is not needed. Python supplies some GDB configuration (https://github.com/python/cpython/blob/master/Tools/gdb/libpython.py), which properly packaged (`apt-get install python3-dbg`) seem to be just working. Maybe this option could be dropped altogether in favour of just `gdb --args graphene -f ...`.

**Comment by mkow** [comment]
+1 to the draft proposal in its current form.

ad. GDB: The command line wouldn't be just `gdb --args graphene -f ...`, we had that switch because running under GDB is quite complicated - for example, we need to LD_PRELOAD a library into GDB to support enclave introspection.

**Comment by chiache** [comment]
Sounds like a reasonable change to me.

I am okay with not adding the `--gdb` argument, but we will need another binary or script to invoke Graphene-SGX in gdb:
```
graphene-sgx-gdb {--manifest|-f} MANIFEST
```
or
```
graphene-sgx-gdb --args graphene-sgx {--manifest|-f} MANIFEST
```

**Comment by woju** [comment]
I'd be vary of having a Cartesian product of `graphene{,-sgx}{,-gdb}`. Could we instead have a single wrapper, maybe like the current `pal_loader`, which would only care about GDB, something like:

```sh
#!/bin/sh

LD_PRELOAD=/usr/lib/graphene/sgxgdbwhatever.so gdb \
  -x /usr/share/graphene/graphene.gdb \
  #whatever else
  --args "$@"
```

So it would be used as

```sh
graphene-gdb graphene[-sgx] [...]
```

"Do one thing and do it well."

Also, as part of the kernel driver upstreaming, there is a patch which fixes problem with `ptrace()`: https://lore.kernel.org/lkml/20190903142655.21943-16-jarkko.sakkinen@linux.intel.com/. This probably will make the LD_PRELOAD obsolete. But this depends on the upstream effort, which I believe has no timeline.

**Comment by donporter** [comment]
I am good with this.  FWIW, I usually just invoke the appropriate PAL directly when I run tests by hand.

**Comment by dimakuv** [comment]
> ### Related work
> - #994 

This made my day :)

I am also good with this. But the comment for option 3 is slightly incorrect? I thought we always require a manifest (but we deduce the manifest from the executable name).

Regarding shebangs, a Perl-like solution seems reasonable.

Regarding GDB, I agree with the latest proposal of @woju: a `graphene-gdb` shell wrapper. In this case, the `--gdb` argument for the main Graphene executable/script is not needed.

**Comment by woju** [comment]
Thanks for the comments. I updated the proposal.

@dimakuv Yes, I wasn't clear in option 3. I meant the explicit pointing to the manifest is not needed, not the data itself. This RFC is more about the external interface (incl. possible future directions), not the inner workings.

**Comment by mkow** [comment]
> I updated the proposal.

LGTM

**Comment by pwmarcz** [comment]
Looks like a good direction. It would unlock proper command-line arguments for Graphene, and also get rid of one layer of indirection.

I'm not sure about the `graphene-gdb` wrapper. There are two GDB integrations (SGX and non-SGX) and they have slightly different ways of running the whole thing. I guess you would have to detect if we're running the SGX version by looking at the argument (`graphene-sgx` or `graphene`, possibly with a path)?

* Apart from gdb, `pal_loader` has a `PERF=` switch that prepends the command with `perf stat`. We can throw that away, right?
* Graphene needs to find the right `libpal-$PAL_HOST.so`. Where do we look for it? `LD_LIBRARY_PATH`? A specific directory?
* How do I run Graphene as a developer? Ideally I would like to run `make` (or `meson build`) and be able to run a new version, without installing anything globally.
  * I might be working on a few versions at the same time (e.g. GCC and clang, or comparing different feature)
  * I might have an automated script that I use with `git bisect` (and I'd rather not use sudo in an automated script)
  * I want to be able to run tests quickly: usually installing a project is not a prerequisite to running tests

  I guess this mostly ties in with the previous question: I want to be able to run a "local" Graphene and use "local" version of libpal.so and other libraries bundled with Graphene.

**Comment by woju** [comment]
> I'm not sure about the graphene-gdb wrapper.

I'm not either. Are there any other projects where we can find some inspiration?

> perf stat. We can throw that away, right?

I believe so.

> Graphene needs to find the right libpal-$PAL_HOST.so. Where do we look for it? LD_LIBRARY_PATH? A specific directory?

I'd check in current directory (relative to the runner), and if not found, then fallback hardcoded or some path known to ld (in the second case we'd have to also install something to `/etc/ld.conf.d`). If hardcoding, `/usr/lib/graphene/` on `*.deb` and `/usr/libexec/graphene` on `*.rpm`.

Then I'd make the meson conversion so that the loader and `.so` would be all in top of the builddir. So the answer to your other question:

> How do I run Graphene as a developer?

...would be:

```sh
meson setup build
ninja -C build
build/graphene-sgx
```

The simplest way without this would be also  `meson --prefix /tmp/graphene` (or `$HOME/tmp/graphene` if you have `noexec`-mounted `/tmp`); then you can `meson install` without sudo.

**Comment by dimakuv** [comment]
Request from me: Could we add "build" flags and choices inside the final executable of Graphene? It would be cool to have something like `graphene-direct --version` which would show the version and maybe month/year. And also `graphene-direct --fullversion` which would also show the set of flags with which Graphene was built.

@woju's comment:
> It's typically done by having some header generated with all the config (like linux' CONFIG_WHATEVER or autotools-style HAVE_ or WITH_), and somewhere inbeetween an #define CONFIG_CFLAGS "..." (it's generated by the config system, and the config system knows the flags). Then you can prinf("%s\n", CONFIG_CFLAGS) to your heart's content. --version probably shouldn't contain all the flags (just graphene-direct X.Y.Z, because it might be parsed), but --fullversion is definitely possible.

**Comment by mkow** [comment]
+1
Also, I'd add this info to the debug output (on "debug" level and above). Will make catching people using old Graphene easier :)

**Comment by woju** [comment]
Note to myself: all these ideas are certainly worth consideration, but some
will require careful implementation to have the build reproducible. Instead of
time of compilation, we will probably have timestamp of the HEAD commit etc.

**Comment by mkow** [comment]
Yup, personally I don't care about the compilation time, but about the commit hash (if compiled from a git repo).

**Comment by mkow** [comment]
One more thing: Please not only include the commit hash, but also build mode (debug/release/...).
